src/outmdsd/DjsonLogItem.cc

```diff
@@ -47,32 +47,16 @@ DjsonLogItem::ComposeSchema(
 {
     IdMgr::value_type_t cachedInfo;
 
-    // because most log items will share same schema, use unsorted key first
-    auto unsortedKey = GetSchemaCacheKey();
-    if (GetIdMgr().GetItem(unsortedKey, cachedInfo)) {
+    // The fields order are preserved. So schema with same names/types
+    // but in different order will be treated different schemas.
+    auto key = GetSchemaCacheKey();
+    if (GetIdMgr().GetItem(key, cachedInfo)) {
         strm << cachedInfo.first << "," << cachedInfo.second;
```

Comment:
Simply using 'null' here won't work, especially when there is a mdsd connection failure. 

Example1, the schema caching occurs at LogItem construction time. so a found-schema doesn't mean the message was sent to mdsd already. It just mean it was constructed before. all these messages could all end up in the backup cache.  However, the current backup cache is not sorted. so when these messages are to be sent to mdsd from the cache, the first construted one may not be the first one to in the cache iteration.

Example2, assuming an app wants to send10 messages with same schema  to mdsd.  The first one with schema str is sent to mdsd properly, then mdsd crashed. The next 9 messages will have 'null' in schema. When mdsd is restarted, the 9 messages don't know their schema info. mdsd won't be able to process them.

Overall, this needs more design. I created TFS to track this: 892965




---
In reply to: [97908578](https://github.com/Azure/fluentd-plugin-mdsd/pull/6#discussion_r97908578) [](ancestors = 97908578)